File: osd/1.2/enterprise.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <sys/wait.h>

void kirk_logs(int sig)
{
    int ch;
    char *path = "/tmp/kirk.log";
    char msg[101] = {0};
    FILE *f = fopen(path,"a+");
    if(NULL != f)
    {
        printf("Kirk, say something:\n");
        fread(msg, 1, 100, stdin);
        fprintf(f,"%s",msg);
        fclose(f);
    }
    else
        printf("No log found\n");

}

int enterprise_to_spock(void)
{
    char *ext_name = "spock";
    //char *ext_name = "nginx";
    int ext_pid;
    FILE *file;
    char *path = "/tmp/spock.pid";

    while(1)
    {
        file = fopen(path,"r");
        if(NULL != file)
        {
            fscanf(file,"%d", &ext_pid);
            printf("Enterprise, %s is at: %d, sending signal...\n", ext_name, ext_pid);
            kill(ext_pid,SIGUSR1);
            fclose(file);
        }
        else
            printf("Enterprise can't find %s\n", ext_name);
    }

    exit(0);
}

int main(void)
{
    pid_t pid;

    char *pidfile = "/tmp/enterprise.pid";
    FILE *pf = fopen(pidfile,"r");
    if(NULL == pf)
    {
        pf = fopen(pidfile,"w");
        fprintf(pf,"%d\n",getpid());
        fclose(pf);
    }
    else
    {
        printf("One USS is enough for our galaxy\n");
        exit(1);
    }

    // Handle SIGUSR1
    signal(SIGUSR1, kirk_logs);

    enterprise_to_spock();

    unlink(pidfile);
    return 0;
}

File: osd/2.2/enterprise.py
#!/usr/bin/env python

import os
import signal
import sys
import time

def kirk_logs(signum, stack):
    logpath = "/tmp/kirk.log"
    print "Kirk, say something:"
    msg = sys.stdin.readline()
    f = open(logpath, "w")
    f.write(repr(msg))
    f.close()

def enterprise_to_spock():
    s_pidfile = "/tmp/spock.py.pid"
    while True:
        if( os.path.exists(s_pidfile) ):
            f = open( s_pidfile )
            s_pid = f.readline()
            f.close()
            print "Enterprise, spock is at:", s_pid, ", sending signal..."
            os.kill(int(s_pid), signal.SIGUSR1)
            time.sleep(10)
        else:
            print "Enterprise can't find spock..."
            time.sleep(10)
    
    sys.exit(0)

def main():
    pidfile = "/tmp/enterprise.py.pid"
    if( os.path.exists(pidfile) ):
        print "One USS is enough for our galaxy. Exiting..."
        sys.exit(1)
    else:
        f = open(pidfile, "w")
        our_pid = os.getpid()
        f.write(repr(our_pid))
        f.close()
    
    signal.signal(signal.SIGUSR1, kirk_logs)
    enterprise_to_spock()
    os.remove(pidfile)
    sys.exit(0)

if __name__ == "__main__":
    main()
